### PR TITLE
Add UI OfferBook list add/remove checks

### PR DESCRIPTION
**Pull request by ghubstan**
Problem:  Using the API's CLI to edit offers causes add/remove messages to be received on peers in the same envelope batch, then passed to the UI in random order.  When `OfferBook` receives a remove(offer) event after an add(offer) event, the edited offer will be removed immediately after being added to the UI's `OfferBook` list.  

Fix:  This change uses storage entry sequence number and storage entry payload hash comparisons to avoid the problem.

- `OfferBookListItem` Added new constructor taking `P2PDataStorage.ByteArray hashOfPayload`,
  and `int sequenceNumber` params.  Defined a `toString` method.
  
- `OfferBook` Added new checks on `OfferBookListItem`'s `hashOfPayload` and `sequenceNumber` while
  determining if offer candidates should be added or removed from the UI's OfferBook List.
  See OfferBook contructor's implementation of OfferBookChangedListener#onAdded and
  OfferBookChangedListener#onRemoved.  Added many (maybe too many) comments explaining the add/remove rules,
  and plenty of debug statements (maybe too many) to help trace the add/remove event process.
  
- `OfferBookService#OfferBookChangedListener` Added new `P2PDataStorage.ByteArray` hashOfPayload,
  and `int sequenceNumber` params to listener's `onAdded` and `onRemoved` method signatures.
  Added these two new parameter values to listener.onAdded and listener.onRemoved calls.
  
- `TakeOfferDataModel` Replaced unused `tradeManager` param in `offerBook.removeOffer`
  with `P2PDataStorage.ByteArray hashOfPayload` (null), and `int sequenceNumber` (-1) params.
  `OfferBook` will remove the candidate offer as before.

- MarketAlerts` Adjusted `onAdded` & `onRemoved` listener method signatures.

This PR is based on https://github.com/bisq-network/bisq/pull/5577, which should be reviewed and merged before this one.

### CODE REVIEW

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java [hunk 1]
```diff
@@ -83,44 +85,138 @@ public void onAdded(Offer offer) {
...
-                    // If we have an offer with same ID we remove it and add the new offer as it might have a changed state.
-                    Optional<OfferBookListItem> candidateWithSameId = offerBookListItems.stream()
-                            .filter(item -> item.getOffer().getId().equals(offer.getId()))
-                            .findAny();
-                    if (candidateWithSameId.isPresent()) {
-                        log.warn("We had an old offer in the list with the same Offer ID. We remove the old one. " +
-                                "old offerBookListItem={}, new offerBookListItem={}", candidateWithSameId.get(), offerBookListItem);
-                        offerBookListItems.remove(candidateWithSameId.get());
+                    OfferBookListItem newOfferBookListItem = new OfferBookListItem(offer, hashOfPayload, sequenceNumber);
+                    removeAnyOldOfferBookListItemsBeforeAddingReplacement(newOfferBookListItem);
```
Comment (sqrrm) [inline desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java]: I find this method name somewhat on the long side

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookListItem.java [hunk 2]
```diff
@@ -79,7 +110,7 @@ public WitnessAgeData getWitnessAgeData(AccountAgeWitnessService accountAgeWitne
                     // either signed & limits lifted, or waiting for limits to be lifted
                     // Or banned
                     daysSinceSignedAsLong = TimeUnit.MILLISECONDS.toDays(optionalWitness.map(witness ->
-                            accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
+                                    accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
```
Comment (sqrrm) [inline desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookListItem.java:113]: Bad indent

**Review by Dpalazue — APPROVED** [review]

**Review by Dpalazue — APPROVED** [review]

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java [hunk 3]
```diff
@@ -229,8 +247,8 @@ public void fillOfferBookListItems() {
             // Investigate why....
             offerBookListItems.clear();
             offerBookListItems.addAll(offerBookService.getOffers().stream()
-                    .filter(o -> !filterManager.isOfferIdBanned(o.getId()))
-                    .filter(o -> !OfferRestrictions.requiresNodeAddressUpdate() || Utils.isV3Address(o.getMakerNodeAddress().getHostName()))
+                    .filter(o -> !isOfferIdBanned(o))
+                    .filter(o -> isV3NodeAddressCompliant(o))
```
Comment (sqrrm) [inline desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java]: If you use `ifOfferAllowed()` both of these filters can use method references instead. I find that's usually preferred when possible.

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java [hunk 4]
```diff
@@ -148,35 +148,52 @@ public void removeOffer(Offer offer, P2PDataStorage.ByteArray hashOfPayload, int
...
+            // The OfferBookListItem with a null or matching payload-hash was not found.
+            // However, when the API's CLI is used to edit and deactivate an offer
+            // in the same command, the edited offer is not re-published (and cannot be
+            // found in local storage).  In this case, we need to remove the deactivated
+            // offer from the list if the local store does not contain an offer with a
+            // matching offerId.
+            if (!isStoredLocally(offer)) {
+                Optional<OfferBookListItem> viewItem = getOfferBookListItem(offer);
+                viewItem.ifPresent((item) -> {
+                    offerBookListItems.remove(item);
```
Comment (sqrrm) [inline desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:179]: Why is this needed? Isn't this just to present it nicely to the UI and that's not a problem when running the cli?

Comment (ghubstan) [reply desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:179]: >> Why is this needed?

Short explanation is a stale offer with no matching offer (with id) in local protected storage cannot be taken and should be removed from view.

>> Isn't this just to present it nicely to the UI and that's not a problem when running the cli?

The `OfferBookListItem` is a JFX observable used for display in the OfferBook views.  This code block supports the CLI being able to atomically edit and deactivate an offer in the same `editoffer` command, something the UI never does.  When you edit an offer with the UI, you deactivate the offer in the 1st mouse click (pencil button), then re-activate it with the confirm-edit or cancel button clicks -- two separate actions that ensure the temporarily deactivated offer is removed from peers, and the edited (or not) offer is re-published to peers (re-activated).

When the CLI edits _and_ deactivates an offer in the same `editoffer` command, the edited&disabled offer _will not be re-published_ by the API.  On another UI/Peer, the old stale OfferBookListItem would remain (without this code block) in the UI's observable item list because the peer receives on `onRemoved` msg containing the payload-hash of the newly edited offer the peer never received;  there is no matching payload-hash, and `OfferBook.onRemoved(offer, payloadHash)` does not know which item to remove.

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookListItem.java (hunk 2 above)
Comment (ghubstan) [reply desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookListItem.java:113]: Either, this is formatting of an older commit (which I have not found yet).  Or my IDE style rules have changed since I upgraded to latest version (doubtful because my bisq style rule file has not changed).

Lines 112-114 in the repo and after auto-formatting in my IDE are
```
                    daysSinceSignedAsLong = TimeUnit.MILLISECONDS.toDays(optionalWitness.map(witness ->
                                    accountAgeWitnessService.getWitnessSignAge(witness, new Date()))
                            .orElse(0L));

```

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java (hunk 3 above)
Comment (ghubstan) [reply desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java]: Requested change in commit d709338d6366294ef727b0ab561e7ba73384872a .

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java (hunk 1 above)
Comment (ghubstan) [reply desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java]: Requested change in commit 59c031327725bfc669414389056b875eff9a9faa.

**Comment by ghubstan** [comment]
File conflicts resolved in child PR https://github.com/bisq-network/bisq/pull/5666

File: desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java (hunk 4 above)
Comment (ghubstan) [reply desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:179]: Commit fb4e00fb6bf6c6fbac536292d3b930f3c04cacf0 fixes the bug that caused me to add this workaround, and it has been removed in PR https://github.com/bisq-network/bisq/pull/5666
